src/oop/object_interactions/medium/q2/MainTest.java

```diff
@@ -0,0 +1,165 @@
+                Arguments.of("John", "John is in love with Alice", new String[]{"Alice", "Bob"}, "Alice", "Alice is in love with John"),
+                Arguments.of("Kim", "Kim has broken her leg", new String[]{"Ali", "John"}, "Ali", "Ali broke Kim's leg")
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("hearSecretInputProvider")
+    public void correctHearSecretMethod(String name, String secret, String[] allFriends, String friend, String friendSecret) throws Throwable {
+        Object[][] arguments = {
+                {name, String.class},
+                {secret, String.class},
+                {allFriends, String[].class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {friend, String.class},
+                {friendSecret, String.class}
+        };
+        Object output = testClass.callMethod(hearSecretMethodName, methodArguments, new String[]{"public"}, classInstance);
+        String[] ans = new String[allFriends.length];
+        for (int i = 0; i < allFriends.length; i++) {
+            if (friend.equals(allFriends[i]))
+                ans[i] = friendSecret;
+        }
+        _assertArrayEquals(ans, testClass.getFieldValue(classInstance, friendSecretsAttributeName),
+                "Your " + hearSecretMethodName + " method does not change the " + friendSecretsAttributeName + " attribute correctly.");
+        assertNull(output, String.join(" ", "Your", hearSecretMethodName, "should not return any output"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("hearSecretInputProvider")
+    public void correctWriteInDiaryMethod(String name, String secret, String[] allFriends, String friend, String friendSecret) throws Throwable {
+        Object[][] arguments = {
+                {name, String.class},
+                {secret, String.class},
+                {allFriends, String[].class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] methodArguments = {
+                {friend, String.class},
+                {friendSecret, String.class}
+        };
+        testClass.callMethod(hearSecretMethodName, methodArguments, new String[]{"public"}, classInstance);
+        String[] ans = new String[allFriends.length];
+        for (int i = 0; i < allFriends.length; i++) {
+            if (friend.equals(allFriends[i]))
+                ans[i] = friendSecret;
+        }
+        Object output = testClass.callMethod(writeInDiaryMethodName, null, new String[]{"public"}, classInstance, new Clause[]{
+                new StringLiteral("I have the following secrets:"),
+                new Optional(new StringLiteral(" ")),
+                new NewLine(),
+                new StringLiteral(friend + " told me " + friendSecret),
+                new Optional(new StringLiteral(".")),
```

Comment:
I'm not sure this should be optional, considering it was part of the sample output